Switch endpoints controller to shared informers

ncdc opened the pull request:
cc @bprashanth @thockin @deads2k @sttts @liggitt @smarterclayton  @kubernetes/sig-scalability-pr-reviews

Commit:
Switch endpoints controller to shared informers

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41114)
<!-- Reviewable:end -->

Commit:
Fix controller manager RBAC in local-up-cluster

Change the name of the client cert used by the controller manager from
system:controller to system:kube-controller-manager, so the appropriate
RBAC rules are applied. Also remove the system:masters group from this
client cert so it doesn't have super powers.

Also, always enable using service account credentials in the controller
manager.

deads2k commented:
`Fix controller manager RBAC in local-up-cluster` lgtm

deads2k commented:
Seems like `Allow controller manager to list/watch Services` should be squashed in here somewhere.

deads2k commented:
The permission it adds is fine though.

ncdc commented:
I can squash the policy change with `Switch endpoints controller to shared informers`

deads2k commented on pkg/controller/endpoint/endpoints_controller.go:
```diff
@@ -338,8 +303,12 @@ func (e *EndpointController) syncService(key string) error {
+	if err != nil {
+		return err
+	}
+	service, err := e.serviceLister.Services(namespace).Get(name)
```
how hard is it to deep copy the map copies that happening?  There's no mutation on `newEndpoints.Labels` as a for instance, but the map is the same as the shared cache for services.

If its non-trivial (it always is in Go), I could live without it.

ncdc replied:
We'd just have to write a for loop to create a new map and copy the KVs in.

deads2k commented:
the controller change lgtm.  I haven't gone back through the resync pull yet.

ncdc commented:
@k8s-bot kubemark e2e test this

timothysc commented:
quick question, why aren't endpoints themselves using informers vs. a hard list + gets.  

/cc @wojtek-t

ncdc commented:
@timothysc I can convert those too

ncdc commented:
I think it makes sense to do the endpoint informer change in a follow-up, as updating the test cases is not trivial.

timothysc commented:
> I think it makes sense to do the endpoint informer change in a follow-up, as updating the test cases is not trivial. 

+1

ncdc commented:
Rebased on top of the latest in #40759 (no other changes)

smarterclayton commented:
Rebase and I'll do final review.

ncdc commented:
Rebased

smarterclayton commented:
I don't see anything obvious.

smarterclayton commented:
/lgtm
/approve

ncdc commented:
Adding the lgtm for Clayton
